pkg/sql/sem/tree/udf.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tree
+
+import (
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/errors"
+)
+
+// FunctionName represent a function name in a UDF relevant statement, either
+// DDL or DML statement. Similar to TableName, it is constructed for incoming
+// SQL queries from an UnresolvedObjectName.
+type FunctionName struct {
+	objName
+}
+
+// Format implements the NodeFormatter interface.
+func (f *FunctionName) Format(ctx *FmtCtx) {
+	f.ObjectNamePrefix.Format(ctx)
+	if f.ExplicitSchema || ctx.alwaysFormatTablePrefix() {
+		ctx.WriteByte('.')
+	}
+	ctx.FormatNode(&f.ObjectName)
+}
+
+// CreateFunction represents a CREATE FUNCTION statement.
+type CreateFunction struct {
+	IsProcedure bool
+	Replace     bool
+	FuncName    FunctionName
+	Args        FuncArgs
+	ReturnType  FuncReturnType
+	Options     FunctionOptions
```

Comment:
Is there or should there be a canonical ordering of these options? I don't think we need to preserve the order the user gave us. In particular, I feel like we may want to put the definition string at the end.

Reply:
Good question. Are you concerned about the formatting of the string representation? Yeah, I think it'd be good to have a canonical ordering of them and having function body at the end. I'll make it that way.

Reply:
Oh.. I actually already put function body at the end of the formatting output.. Maybe you have different concerns. One thing to note that we'll convert this statement into a PlanNode which would have these options as individual fields, and we'd add validation on conflicting options like having `IMMUTABLE` and `STABLE` in one statement...